File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.cpp
class Solution {
public:
    vector<string> res;
    vector<int> used;
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        sort(tickets.begin(), tickets.end());
        used = vector<int>(tickets.size(), 0);
        res.push_back("JFK");
        backtracking(tickets);
        return res;
    }

    bool backtracking(vector<vector<string>>& tickets) {
        if (res.size() == tickets.size() + 1) {
            return true;
        }

        for (int i = 0; i < tickets.size(); i++) {
            if (tickets[i][0] == res[res.size() - 1] && used[i] == 0) {
                res.push_back(tickets[i][1]);
                used[i] = 1;
                if (backtracking(tickets)) {
                    return true;
                }
                used[i] = 0;
                res.pop_back();
            }
        }

        return false;
    }
};

File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.py
class Solution:
  
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        graph = defaultdict(list)

        for u, v in sorted(tickets):
            graph[u].append(v)

        res = []

        def dfs(curr):
            while graph[curr]:
                next_city = graph[curr].pop(0)
                dfs(next_city)

            res.append(curr)

        dfs("JFK")

        res.reverse()
        return res
             